camdecmpswks/functions/rpt_em_emission_view_summary.sql

```diff
@@ -10,38 +10,92 @@ CREATE OR REPLACE FUNCTION camdecmpswks.rpt_em_emission_view_summary(
 
 RETURNS TABLE (
     "period_description" varchar,
-    "row_name" text,
-    "op_hours" numeric,
-    "op_time" numeric,
-    "heat_input" numeric,
-    "so2_mass" numeric,
-    "co2_mass" numeric,
-    "nox_rate" numeric,
-    "nox_mass" numeric
-)
+    "op_hours" NUMERIC,
+    "op_time" NUMERIC,
+    "heat_input" NUMERIC,
+    "so2_mass" NUMERIC,
+    "co2_mass" NUMERIC,
+    "nox_rate" NUMERIC,
+    "nox_mass" NUMERIC
+) 
 LANGUAGE 'plpgsql'
 COST 100
 VOLATILE
 ROWS 1000
 
 AS $BODY$
+
+DECLARE
+    year NUMERIC;
+    maxQuarter NUMERIC;
 BEGIN
+    FOR year IN
+        SELECT DISTINCT rp.calendar_year 
+        FROM camdecmpsmd.reporting_period rp
+        WHERE rp.rpt_period_id IN (SELECT unnest(string_to_array(reportingPeriodIds, ',')::NUMERIC[]))
+        ORDER BY rp.calendar_year
+    LOOP
+        -- Get the last quarter for this year from the passed-in reporting period id list
+        SELECT MAX(rp.rpt_period_id) INTO maxQuarter
+        FROM camdecmpsmd.reporting_period rp
+        WHERE rp.calendar_year = year
+          AND rp.rpt_period_id IN (SELECT unnest(string_to_array(reportingPeriodIds, ',')::NUMERIC[]));
+
+        RETURN QUERY
+        -- get data for current quarter and all previous quarters for the calender year
+        SELECT evs.period_description,
+               evs.op_hours,
+               evs.op_time,
+               evs.heat_input,
+               evs.so2_mass,
+               evs.co2_mass,
+               evs.nox_rate,
+               evs.nox_mass
+        FROM camdecmpswks.emission_view_sumval evs
+            JOIN camdecmpsmd.reporting_period rp USING(rpt_period_id)
+        WHERE evs.mon_plan_id = monitorPlanId
+          AND evs.mon_loc_id = locationId
+          AND evs.rpt_period_id <= maxQuarter
```

Comment:
Generally we do not use RPT_PERIOD_ID for comparing or ordering.  Generally, you should not use a primary key in this way.

In this case you can use a Max on QUARTER in the select at line 39, and compare rp.quarter instead of evs.rpt_period_id in this line (58).  That new comparison in conjunction with the condition "rp.calendar_year = year" should have the same affect.